templates/common/upgrade/hash.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package upgrade
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/abcxyz/abc/templates/common"
+)
+
+// TODO(upgrade): delete this once the rest of the upgrade code exists. The
+// compiler is complaning about unused variables.
+var ExportToAvoidWarnings = []any{hashAndCompare, notComputed}
+
+// hashResult is the result from hashAndCompare().
+type hashResult string
+
+const (
+	// notComputed is the zero value, meaning "the hash was not computed
+	// (yet)".
+	notComputed hashResult = ""
+
+	// match means "the file contents were hashed, and the value of the hash
+	// matched the expected value."
+	match hashResult = "hash_matched"
+
+	// mismatch means "the file contents were hashed, and the value of the
+	// hash didn't match the expected value."
+	mismatch hashResult = "edited"
+
+	// deleted means "the file contents couldn't be hashed because the file
+	// doesn't exist."
+	deleted hashResult = "deleted"
+)
+
+// hashAndCompare extracts the hash algorithm (e.g. "h1:" from wantHash, then
+// hashes the given path with that algorithm.
+func hashAndCompare(path string, wantHash string) (hashResult, error) {
+	// The hash should start with a string like "h1:" indicating the hash algorithm
+	tokens := strings.SplitN(wantHash, ":", 2)
+	if len(tokens) != 2 {
+		return "", fmt.Errorf("malformed hash, expected it to begin with hash name followed by colon: %q", wantHash)
+	}
+
+	var hasher hash.Hash
+	var wantHashUnmarshaled []byte
+	switch tokens[0] {
+	case "h1":
+		hasher = sha256.New()
```

Comment:
From an extensibility perspective, it might be better if we let the user supply a hashing function (as a string) and have a default on `""`. We can have a `switch` on the string among common hashing functions (md5, sha128, sha256, sha384, sha512, etc).

Reply:
I'm doing this for consistency with dirhash, which we already use [here](https://github.com/abcxyz/abc/blob/ca35eed9cc7b1fa23801a84b877d2738c32a58e8/templates/common/dirhash/dirhash.go#L40). This way, we can have "h1" mean "use sha256 for dirhashing and also for individual file hashes." I don't think we want to give the user any control here, we want to limit choice for the sake of simplicity.

Here's an example of a manifest:

```
# Generated by the "abc templates" command. Do not modify.
api_version: cli.abcxyz.dev/v1beta6
kind: Manifest
creation_time: 2023-12-08T23:59:02.000000013Z
modification_time: 2023-12-08T23:59:02.000000013Z
template_location: "github.com/foo/bar"
location_type: "remote_git"
template_version: ""
template_dirhash: h1:uh/nUYc3HpipWEon9kYOsvSrEadfu8Q9TdfBuHcnF3o=
inputs: []
output_hashes:
    - file: a.txt
      hash: h1:ZmFrZV9vdXRwdXRfaGFzaF8zMl9ieXRlc19zaGEyNTY=
```

I think the there's nice conceptual simplicity in using the same "hash framing" scheme for both dirhashes and file hashes.

Reply:
Ah - so you don't plan to expose this as a template func?

Reply:
Right, yeah, this is something totally else. That would be a neat template func though.